Fix flaky camera stream teardown

ptarjan opened the pull request:
Mock create_stream instead of Stream.endpoint_url to prevent real stream objects from being created during the test. This fixes a flaky test failure where the test would fail with "errors while tearing down" because the stream worker thread wasn't being properly cleaned up when the test patches exited.

This failed in https://github.com/home-assistant/core/actions/runs/20080972481/job/57608219715?pr=158279

<!--
  You are amazing! Thanks for contributing to our project!
  Please, DO NOT DELETE ANY TEXT from this template! (unless instructed).
-->
## Breaking change
<!--
  If your PR contains a breaking change for existing users, it is important
  to tell them what breaks, how to make it work again and why we did this.
  This piece of text is published with the release notes, so it helps if you
  write it towards our users, not us.
  Note: Remove this section if this PR is NOT a breaking change.
-->


## Proposed change
<!--
  Describe the big picture of your changes here to communicate to the
  maintainers why we should accept this pull request. If it fixes a bug
  or resolves a feature request, be sure to link to that issue in the
  additional information section.
-->


## Type of change
<!--
  What type of change does your PR introduce to Home Assistant?
  NOTE: Please, check only 1! box!
  If your PR requires multiple boxes to be checked, you'll most likely need to
  split it into multiple PRs. This makes things easier and faster to code review.
-->

- [ ] Dependency upgrade
- [x] Bugfix (non-breaking change which fixes an issue)
- [ ] New integration (thank you!)
- [ ] New feature (which adds functionality to an existing integration)
- [ ] Deprecation (breaking change to happen in the future)
- [ ] Breaking change (fix/feature causing existing functionality to break)
- [ ] Code quality improvements to existing code or addition of tests

## Additional information
<!--
  Details are important, and help maintainers processing your PR.
  Please be sure to fill out additional details, if applicable.
-->

- This PR fixes or closes issue: fixes #
- This PR is related to issue: 
- Link to documentation pull request: 
- Link to developer documentation pull request: 
- Link to frontend pull request: 

## Checklist
<!--
  Put an `x` in the boxes that apply. You can also fill these out after
  creating the PR. If you're unsure about any of them, don't hesitate to ask.
  We're here to help! This is simply a reminder of what we are going to look
  for before merging your code.

  AI tools are welcome, but contributors are responsible for *fully*
  understanding the code before submitting a PR.
-->

- [x] I understand the code I am submitting and can explain how it works.
- [x] The code change is tested and works locally.
- [x] Local tests pass. **Your PR cannot be merged unless tests pass**
- [x] There is no commented out code in this PR.
- [x] I have followed the [development checklist][dev-checklist]
- [x] I have followed the [perfect PR recommendations][perfect-pr]
- [x] The code has been formatted using Ruff (`ruff format homeassistant tests`)
- [x] Tests have been added to verify that the new code works.
- [x] Any generated code has been carefully reviewed for correctness and compliance with project standards.

If user exposed functionality or configuration variables are added/changed:

- [ ] Documentation added/updated for [www.home-assistant.io][docs-repository]

If the code communicates with devices, web services, or third-party tools:

- [ ] The [manifest file][manifest-docs] has all fields filled out correctly.  
      Updated and included derived files by running: `python3 -m script.hassfest`.
- [ ] New or updated dependencies have been added to `requirements_all.txt`.  
      Updated by running `python3 -m script.gen_requirements_all`.
- [ ] For the updated dependencies - a link to the changelog, or at minimum a diff between library versions is added to the PR description.

<!--
  This project is very active and we have a high turnover of pull requests.

  Unfortunately, the number of incoming pull requests is higher than what our
  reviewers can review and merge so there is a long backlog of pull requests
  waiting for review. You can help here!
  
  By reviewing another pull request, you will help raise the code quality of
  that pull request and the final review will be faster. This way the general
  pace of pull request reviews will go up and your wait time will go down.
  
  When picking a pull request to review, try to choose one that hasn't yet
  been reviewed.

  Thanks for helping out!
-->

To help with the load of incoming pull requests:

- [ ] I have reviewed two other [open pull requests][prs] in this repository.

[prs]: https://github.com/home-assistant/core/pulls?q=is%3Aopen+is%3Apr+-author%3A%40me+-draft%3Atrue+-label%3Awaiting-for-upstream+sort%3Acreated-desc+review%3Anone+-status%3Afailure

<!--
  Thank you for contributing <3

  Below, some useful links you could explore:
-->
[dev-checklist]: https://developers.home-assistant.io/docs/development_checklist/
[manifest-docs]: https://developers.home-assistant.io/docs/creating_integration_manifest/
[quality-scale]: https://developers.home-assistant.io/docs/integration_quality_scale_index/
[docs-repository]: https://github.com/home-assistant/home-assistant.io
[perfect-pr]: https://developers.home-assistant.io/docs/review-process/#creating-the-perfect-pr

ptarjan committed:
Fix flaky test_websocket_camera_stream teardown

Mock create_stream instead of Stream.endpoint_url to prevent real stream
objects from being created during the test. This fixes a flaky test failure
where the test would fail with "errors while tearing down" because the
stream worker thread wasn't being properly cleaned up when the test
patches exited.

epenet reviewed (approved):
LGTM 👍

ptarjan committed:
Apply suggestions from code review

Co-authored-by: epenet <6771947+epenet@users.noreply.github.com>

ptarjan committed:
Fix ruff errors: define mock_stream before use in patch

Move mock_stream creation before the with block so it can be used
as return_value in the patch() call.

epenet reviewed (approved):
LGTM 👍

ptarjan committed:
Add mock_create_stream fixture for camera tests

Create a reusable fixture that mocks create_stream to prevent real
stream threads from being created during tests. This fixes flaky
teardown errors in test_websocket_camera_stream and
test_handle_play_stream_service.

The fixture provides a mock stream with:
- add_provider: Mock for adding stream providers
- start: AsyncMock for async stream start
- endpoint_url: Returns a mock URL for stream endpoint

ptarjan committed:
Apply suggestions from code review

Co-authored-by: epenet <6771947+epenet@users.noreply.github.com>

ptarjan committed:
Fix test_stream_unavailable teardown by using mock_create_stream fixture

Add set_update_callback and available attributes to the mock_create_stream
fixture to support tests that need to simulate stream availability changes.
Updated test_stream_unavailable to use the fixture instead of patching
Stream methods directly.

epenet reviewed (commented):
Looks good for the two tests.
Can we also update `test_no_preload_stream` and `test_stream_unavailable` to match?

ptarjan committed:
Update test_no_preload_stream and test_preload_stream to use fixture

Use mock_create_stream fixture consistently across preload stream tests
to prevent real stream threads from being created during tests.

epenet reviewed (commented):
A couple of comments, then I think we're good
